### PR TITLE
Add unlit shading to `ModelExperimental`

### PR DESCRIPTION
**Pull request by ptrgags**
This branch is targeting a branch, not `main`.

Opening it as a draft PR since I see some documentation and test changes that are better handled in `model-experimental` rather than this branch.

As part of the Model.js refactor, this builds on our base code and adds:

* a `MaterialPipelineStage` which processes glTF materials and adds them as uniforms to the shader. This adds shader code to pre-process the materials, but does not do any lighting.
* a `LightingPipelineStage` goes at the end of the pipeline, takes a material and applies the appropriate lighting model (right now only unlit, a future PR will add PBR lighting)

@sanjeetsuhag could you review?

CC @lilleyse

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ptrgags!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Review by lilleyse — COMMENTED** [review]
@ptrgags looks great so far. While I didn't review everything in fine detail the shader code seems to check out.

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+vec4 SRGBtoLINEAR4(vec4 srgbIn) 
+{
+    vec3 linearOut = pow(srgbIn.rgb, vec3(2.2));
+    return vec4(linearOut, srgbIn.a);
+}
+
+#ifdef HAS_NORMALS
+vec3 computeNormal() {
+  vec3 ng = normalize(v_normal);
+  vec3 positionWC = vec3(czm_inverseView * vec4(v_positionEC, 1.0));
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Source/Shaders/ModelExperimental/MaterialStageFS.glsl:25]: Oh I see, looking back to `processPbrMaterials.js` this was only used for procedural IBL lighting. I'll remove this for now, and we can look at this again on the PBR branch

File: Source/Shaders/Builtin/Structs/modelMaterial.glsl [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * @name czm_modelMaterial
+ * @glslStruct
+ *
+ * @property {vec3} diffuse Incoming light that scatters evenly in all directions.
+ * @property {float} alpha Opacity of this material. 0.0 is completely transparent; 1.0 is completely opaque.
+ * @property {vec3} specular Color of reflected light at normal incidence in PBR materials. This is sometimes referred to as f0 in the literature.
+ * @property {float} roughness A number from 0.0 to 1.0 representing how rough the surface is. Values near 0.0 produce glossy surfaces, while values near 1.0 produce rough surfaces.
+ * @property {vec3} normal Surface's normal in eye coordinates. It is used for effects such as normal mapping. The default is the surface's unmodified normal.
+ * @property {vec3} emission Light emitted by the material equally in all directions. The default is vec3(0.0), which emits no light.
+ * @property {float} alpha Opacity of this material. 0.0 is completely transparent; 1.0 is completely opaque.
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Source/Shaders/Builtin/Structs/modelMaterial.glsl:16]: This is what the old czm_material used. I'll change the wording in both places.

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+  material.roughness = 0.0;
+  material.occlusion = 1.0;
+  material.emissive = vec3(0.0);
+  return material;
+}
+
+vec3 SRGBtoLINEAR3(vec3 srgbIn) 
+{
+    return pow(srgbIn, vec3(2.2));
+}
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Source/Shaders/ModelExperimental/MaterialStageFS.glsl:14]: The first one isn't supported in Edge and the second one isn't supported in Safari. Neither are supported in IE

[the remark this answers is not in the document]
Comment (ptrgags) [reply Source/Shaders/ModelExperimental/MaterialStageFS.glsl:14]: I'm going to leave it this way for now, but i'll make a note to discuss this with you when you're back

File: Source/Shaders/ModelExperimental/ModelExperimentalFS.glsl [hunk 4]
```diff
@@ -1,4 +1,23 @@
...
+    return vec4(color, alpha);
+    #else // OPAQUE
+    return vec4(color, 1.0);
+    #endif
+}
+
 void main() 
 {
-    gl_FragColor = vec4(0.0, 0.0, 1.0, 1.0);
+  czm_modelMaterial material = defaultModelMaterial();
```
Comment (ptrgags) [inline Source/Shaders/ModelExperimental/ModelExperimentalFS.glsl:17]: Indentation should be 4 spaces

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl [hunk 5]
```diff
@@ -0,0 +1,214 @@
+czm_modelMaterial defaultModelMaterial() {
```
Comment (ptrgags) [inline Source/Shaders/ModelExperimental/MaterialStageFS.glsl:1]: Update the formatting on this file. 4 space tabs, K&C style braces based on our style guide.

File: Source/Shaders/ModelExperimental/LightingStageFS.glsl [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    #ifndef HDR 
+    return pow(linearIn, vec3(1.0/2.2));
+    #else 
+    return linearIn;
+    #endif 
+}
+
+czm_modelMaterial lightingStage(czm_modelMaterial inputMaterial)
+{
+    // copying so normal
```
Comment (ptrgags) [inline Source/Shaders/ModelExperimental/LightingStageFS.glsl:12]: finish this comment

File: Source/Scene/ModelExperimental/ModelExperimentalSceneGraph.js [hunk 7]
```diff
@@ -111,14 +111,17 @@ function initialize(sceneGraph) {
     sceneGraph._forwardAxis
   );
 
-  var rootNode = sceneGraph._modelComponents.scene.nodes[0];
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/ModelExperimentalSceneGraph.js:114]: This set of changes should go in #9708 so this can be unit tested along with everything else.

File: Source/Scene/ModelExperimental/LightingPipelineStage.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import ShaderDestination from "../../Renderer/ShaderDestination.js";
+import LightingStageFS from "../../Shaders/ModelExperimental/LightingStageFS.js";
+
+/**
+ * The lighting pipeline stage is responsible for taking a material and rendering
+ * it with a lighting model such as physically based rendering (PBR) or unlit
+ * shading
+ * @private
+ */
+export default function LightingPipelineStage() {}
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/LightingPipelineStage.js:10]: This should be a namespace, not a function. same applies to #9708.

File: Source/Scene/ModelExperimental/MaterialPipelineStage.js [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+ * to render a primitive. This handles the following material types:
+ * <ul>
+ *   <li>Basic glTF materials (PBR metallic roughness model)</li>
+ *   <li>The `KHR_materials_pbrSpecularGlossiness` glTF extension</li>
+ *   <li>The `KHR_materials_unlit` glTF extension</li>
+ * </ul>
+ *
+ * @private
+ */
+export default function MaterialPipelineStage() {}
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/MaterialPipelineStage.js:18]: Same here, convert to a namespace

File: Source/Scene/ModelExperimental/LightingPipelineStage.js (hunk 8 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/LightingPipelineStage.js:10]: (the reason being that this will avoid losing points in code coverage)

File: Source/Scene/ModelExperimental/MaterialPipelineStage.js [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+      ShaderDestination.FRAGMENT
+    );
+    shaderBuilder.addDefine(
+      "TEXCOORD_EMISSIVE",
+      "v_texCoord_" + texCoordIndex,
+      ShaderDestination.FRAGMENT
+    );
+
+    if (defined(emissiveTexture.transform)) {
+      shaderBuilder.addDefine(
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/MaterialPipelineStage.js:90]: I need to find (or make) a model with texture transforms for emissive, normal, etc.

**Review by ptrgags — COMMENTED** [review]
First pass self-review. 

@sanjeetsuhag I discovered some things in #9708 that should be addressed, I'll update your branch

File: Source/Scene/ModelExperimental/MaterialPipelineStage.js (hunk 10 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/MaterialPipelineStage.js]: Nevermind, when I refactored the texture processing code, I added some unit tests that cover the texture transform processing.

**Comment by ptrgags** [comment]
@sanjeetsuhag updated! though let's get #9708 across the finish line before focusing on this

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+            roughness *= u_roughnessFactor;
+            #endif
+        #else
+            #if defined(HAS_METALLIC_FACTOR)
+            float metalness = clamp(u_metallicFactor, 0.0, 1.0);
+            #else
+            float metalness = 1.0;
+            #endif
+
+            #if defined(USE_ROUGHNESS_FACTOR)
```
Comment (ptrgags) [inline Source/Shaders/ModelExperimental/MaterialStageFS.glsl:200]:
```suggestion
            #if defined(HAS_ROUGHNESS_FACTOR)
```

**Comment by ptrgags** [comment]
@sanjeetsuhag this is ready for review, looks like the test failure on Friday was just Travis being slow

File: Source/Scene/ModelExperimental/LightingModel.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
+/**
+ * Which lighting model to use for lighting a {@link ModelExperimental}. This
```
Comment (sanjeetsuhag) [inline Source/Scene/ModelExperimental/LightingModel.js:2]:
```suggestion
 * The lighting model to use for lighting a {@link ModelExperimental}. This
```

File: Source/Scene/ModelExperimental/LightingPipelineStage.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+var LightingPipelineStage = {};
+
+/**
+ * Process a primitive. This modifies the following parts of the render
+ * resources:
+ * <ul>
+ *   <li>modifies the shader to include the lighting stage</li>
+ * </ul>
+ * @param {PrimitiveRenderResources} renderResources The render resources for the primitive
+ * @param {ModelComponents.Primitive} primitive The primitive to be rendered. Not used, but present for consistency with other pipeline stages.
```
Comment (sanjeetsuhag) [inline Source/Scene/ModelExperimental/LightingPipelineStage.js:20]: For the `InstancingPipelineStage`, we need to pass in the `frameState` too. Do we always mention in the documentation when a parameter is unused?

File: Source/Shaders/Builtin/Structs/modelMaterial.glsl [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+ * Struct for representing a material for a {@link ModelExperimental}. The model
+ * rendering pipeline will pass this struct between material, custom shaders,
+ * and lighting stages. This is not to be confused with {@link czm_material}
+ * which is used by the older Fabric materials system, although they are similar
+ *
+ * @name czm_modelMaterial
+ * @glslStruct
+ *
+ * @property {vec3} diffuse Incoming light that scatters evenly in all directions.
+ * @property {float} alpha Opacity of this material. 0.0 is completely transparent; 1.0 is completely opaque.
```
Comment (sanjeetsuhag) [inline Source/Shaders/Builtin/Structs/modelMaterial.glsl:11]:
```suggestion
 * @property {float} alpha Alpha of this material. 0.0 is completely transparent; 1.0 is completely opaque.
```

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+
+#ifdef HAS_NORMALS
+vec3 computeNormal()
+{
+    vec3 ng = normalize(v_normal);
+
+    vec3 normal = ng;
+    #ifdef HAS_NORMAL_TEXTURE
+        vec2 normalTexCoords = TEXCOORD_NORMAL;
+            #ifdef HAS_NORMAL_TEXTURE_TRANSFORM
```
Comment (sanjeetsuhag) [inline Source/Shaders/ModelExperimental/MaterialStageFS.glsl:31]: Why is this indented further in and the one below not?

File: Source/Shaders/ModelExperimental/ModelExperimentalFS.glsl [hunk 16]
```diff
@@ -1,4 +1,23 @@
...
+
 void main() 
 {
-    gl_FragColor = vec4(0.0, 0.0, 1.0, 1.0);
+    czm_modelMaterial material = defaultModelMaterial();
+
+    material = materialStage(material);
+    material = lightingStage(material);
+
+    gl_FragColor = handleAlpha(material.diffuse, material.alpha);
```
Comment (sanjeetsuhag) [inline Source/Shaders/ModelExperimental/ModelExperimentalFS.glsl:22]: Call the function here on a separate line for clarity.

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl (hunk 15 above)
Comment (ptrgags) [reply Source/Shaders/ModelExperimental/MaterialStageFS.glsl:31]: ah good catch, this one was incorrect.

I really don't know the best way to indent code with `#ifdef`s in a readable way. I've been going with the rule "if I nest an `#ifdef`, indent that block 1 more level"

File: Source/Scene/ModelExperimental/LightingPipelineStage.js (hunk 13 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/LightingPipelineStage.js:20]: Hm, this is a good question. Though in this case, Instancing is a different type of pipeline stage because it takes a NodeRenderResources, right?

:open_mouth:  oh this should be `RenderResources.PrimitiveRenderResources` I'll fix that much for now.

**Comment by ptrgags** [comment]
@sanjeetsuhag updated everything except that one part about documenting the pipeline stages which I think we should discuss more

**Comment by sanjeetsuhag** [comment]
@ptrgags Here's 3 problems that need to be fixed before merging:

- [ ] Add support for `incrementallyLoadTextures` in `ModelExperimental`. By default, this should be `true` for models and `false` for 3D Tiles. This is currently causing a race condition that results in a crash. [Local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=rVPbbtpAEP2VlZ9AgrW5NKGEoEo0UKmJGhXUvvhlscd41fWuNbuG0Cr/3vEFhxIeK/lh53Lmcs54AVYWGX94yQFlBtoJtQThCgTLQYutgicTgzqPs3vmsIC7UId6L5DtJRwAyavhwBZ1vR+VrxN6UWUvDAGlBgy9bovLjZVOGk3IBrUQ6Ogl9IgnaLLPsEMA2wk1Y/3BcMSD2/H4ZvCxVzrGYx58CEa3wU1lBqEuK5d10/ximC8gYql3z9JF6XejVCfosfJ7GyWRLxAvUWSwQaFtYjB7m6p1Wa5MRPxUeWbZYlZAmwlnsJo09LRBl4Zer7YOYF3o1ePV3bKSUapfE8dtRHieE71Ex554F3FcVWr6vxOgImelXNL5U6YxVqCaUqfUudxOfR/Fge+kS4ttYQEj4p5gPDKZ/zVFo41doSlyf6c2y/5aZLmCftXD+pmwDtAf8sB/FARDXWWdDL6jps1irF7jSTiUL9MrXKUXrJ8z1qkrsPYGeicHide+m5oPSsncGhnzn6v1ZNyGr4hWh7qnAaWWWZE9U55ay98wZYPhpIq9dk96/JcjnDRHOAhOZ1grG9FsKHiijhtTaxXTMUgtyo7Tf5ePC2zcVOO1e+f1vJl1RwXz076fZJbTZZVqdzj3HZBygkb1t0X0C0hga8veZerMP4fOYrlnMr6/8jeySAlrKZIUquIo9OYzn/LfQZWp5Py2B1TiWKalg/lj7eScz3wyryOdMWor8KLyXw) using [this sample model](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Lantern/glTF).
  image


- [ ] Look into fragment shader compilation issues with `KHR_texture_transform`. [Local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=rVPbbtpAEP2VlZ9AgrW5NKGEoEo0UKmJmhbUvvhlscd41fWuNbvmkir/3vEFQgmPlfywczlzOWc8AyuLjD/sc0CZgXZCzUG4AsFy0GKt4MnEoM7j7J45LOAu1KHeCmRbCTtA8mrYsVld72fla4VeVNkzQ0CpAUOvfcLlxkonjSZkg5oJdPQSesATNNln2CCAbYWasW6vP+DB7XB40/vYKR3DIQ8+BIPb4KYyg1CXlcu6aX4xzBcQsdSbZ+mi9IdRqhV0WPm9jZLIPcRzFBmsUGibGMzepjq5LFcmIn6qPDM/YRZAmwlnsJo09LRBl4Zep7Z2YF3o1ePV3bKSUapfE8dtRHieE71Ex5Z4F3FcVWr6vxOgImehXNL6U6YxVqAaU6fUudyOfR/Fjm+kS4t1YQEj4p5gPDKZ/zVFo41doClyf6NW8+5SZLmCbtXD+pmwDtDv88B/FARDXWd9L8iSLxAf3XxD7ZsVWb3Qk3Ao9+MrrKUX/J9z16orsNM1dI4OkvH0bmo+KCVza2TMfy2Wo+EpfEW+OtQ+Dii1zIrsmfLUkvYYs15/VMVe20dl/ss5jppz7AXHg6w1jmg2FDxRh5WpVYvpLKQWZcfxv8vHBTZuqvHavvM63sS6g4Lpcd9PMsvpxkrdW5z7DkhDQaP66yL6DSS1tWXvMnXin0MnsdwyGd9f+S9ZpIS1FEkKVXEUetOJT/nvoMpUcn7bAipxKNPS3vSxdnLOJz6Z15HOGLUWeFH5Lw) using [this sample model](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Lantern/glTF-Quantized).
  image

- [ ] Intermittent test failures on CI with `MaterialPipelineStageSpec`
  image

**Comment by ptrgags** [comment]
@sanjeetsuhag actually there was a 4th error we found earlier today, 

- [ ] in [this sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=bVJRb9owEP4rVp6CxJxQWNtRiqbSwh5aDZVue8mLSS5gzbEj26Gwqf99ZxtSRpH8YN/d991333kChjcVfdjWoHkF0jIxBWYbDYaCZEsBT6oAcZwnt8TqBm4ymckN02TD4RU0RiW8kkng++ljcRbl/j1RCOQSdBZ1WlytDLdcSUTuUROmLd6Y7NNSq+oeVhrAxJkk5FPvok/Tq8Hgsvel6wKDAU0/p/2r9NI/00w6Zse7rk/EfANWcLmac5uvn5UQcdol7rxLKfkWiqlmFbxoJk2pdPWuqg0ZKlSO/vg6NW0xM8DJmFXaK80iqbRdZ1E3vF7B2CwK8kK3yjmK/ME4anLE0xrtRTs26DsrCs+07/9hAd6cmbBl/NeVEdJoMcROyaKG3CT3zLLEg0wyv3tO7tT2hxTcthe6QuxeHwlqnpjVfDs8M/L6xLzjwePAQNpVdg8B3EF733M+CMFro3hBf80W14M2fcb7kOocBHLJq6aaY51Y8D8wJL2La59767S2BitzZNGMlmL3ooI5BbrPJXPihv/LLBq9D+PfeevcRN1oZOxOwPig7CuvalylszemNLFQ1YLhB02WTf4bLM2Ncd1d6Sg5ho4KviG8uD3z/UkumDGYKRvhp8mi8SjB+g9Qobzx3zegBdu5snVv/BiClNJRgs/zSKuUWDJ9wvwP) the green cube doesn't show up. I've checked the model matrix, bounding sphere, base color, render state. Still not sure why it's not visible.

**Comment by ptrgags** [comment]
@sanjeetsuhag More bugs:

- [x] In `ModelExperimentalSceneGraphSpec.js` the test with a model with vertex colors was failing due to inconsistent usage of `v_vertexColor` and `v_color_0`. I changed everything to use the latter.
- [ ] in `ModelExperimentalSceneGraphSpec.js` it seems the draw command is duplicated somehow 
image

Not sure why this is.

**Comment by ptrgags** [comment]
Fixed the duplicated draw command, I think what was happening is the loading promise was calling `model.update()` and then the test called `model.update()` as well, resulting in two draw commands. Clearing the list up front seems to fix that part at least.

**Comment by ptrgags** [comment]
I've been trying out other glTF models from `/Specs/Data/Models` as part of investigating the missing cube. This is not the only model with issuess, `BoomBox` and `BoomBoxSpecularGlossiness` have the same issue.

I also encountered a couple other things that need addressing:

- [ ] VertexColorTest isn't assigning the vertex colors, the quad renders as solid black instead of half red/half blue
- [ ] The PBR material affects the diffuse color due to the `metallicFactor` and reflectance value. This shouldn't happen for unlit lighting. I think I need to insert some more `#ifdefs`

**Comment by ptrgags** [comment]
@sanjeetsuhag okay at this point I've addressed all the bugs that have come up. I'm still not sure why the timeouts are happening, I also am having difficulty reproducing. Locally I get timeouts, but in an unrelated part of the code (CesiumViewer for example). Furthermore, it's weird that the branch build succeeds but the PR build fails, when I've double-checked that this branch is up-to-date with the latest `model-experimental-staging`

File: Source/Scene/GltfLoader.js [hunk 17]
```diff
@@ -615,15 +615,17 @@ function loadTexture(
...
+    .then(function (textureLoader) {
+      if (loader.isDestroyed()) {
+        return;
+      }
+      textureReader.texture = textureLoader.texture;
+      if (defined(samplerOverride)) {
+        textureReader.texture.sampler = samplerOverride;
+      }
+    })
+    .otherwise(console.error);
```
Comment (sanjeetsuhag) [inline Source/Scene/GltfLoader.js:628]: Look at how `gltfJsonLoader.promise` handles the `otherwise` section.

File: Source/Scene/ModelExperimental/MaterialPipelineStage.js [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+    lightingOptions.lightingModel = LightingModel.UNLIT;
+  } else {
+    lightingOptions.lightingModel = LightingModel.PBR;
+  }
+
+  // Configure back-face culling
+  var cull = !material.doubleSided;
+  renderResources.cull = cull;
+  renderResources.renderStateOptions.cull = {
+    enabled: !material.doubleSided,
```
Comment (sanjeetsuhag) [inline Source/Scene/ModelExperimental/MaterialPipelineStage.js:74]:
```suggestion
    enabled: cull
```

File: Specs/Data/Models/PBR/BoxUnlit/BoxUnlit-processed.gltf [hunk 19]
```diff
@@ -0,0 +1,160 @@
+{
```
Comment (sanjeetsuhag) [inline Specs/Data/Models/PBR/BoxUnlit/BoxUnlit-processed.gltf:1]: Why is this file needed? Can we just modify the original BoxUnlit?

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+
+    #ifdef HAS_OCCLUSION_TEXTURE
+    vec2 occlusionTexCoords = TEXCOORD_OCCLUSION;
+        #ifdef HAS_OCCLUSION_TEXTURE_TRANSFORM
+        occlusionTexCoords = computeTextureTransform(occlusionTexCoords, u_occlusionTextureTransform);
+        #endif
+    material.occlusion = texture2D(u_occlusionTexture, occlusionTexCoords).r;
+    #endif
+
+    #if defined(HAS_EMISSIVE_TEXTURE)
```
Comment (sanjeetsuhag) [inline Source/Shaders/ModelExperimental/MaterialStageFS.glsl:109]: Why not stick to #ifdef when there's only one condition?

**Comment by sanjeetsuhag** [comment]
|Expected Result|Actual Result|
|-|-|
|Screen Shot 2021-08-11 at 9 58 12 AM|Screen Shot 2021-08-11 at 9 58 08 AM|

Select BoxEmissive in this [Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=xVZdb9MwFP0rVp8yqU06Vr5GN8HGNiZRMdHCHggPXnLbWjh2ZDtZBtp/59pO0mgE0AYFqU3ia997zrnOxympIiWDa1DkgAi4JsegWZGFH10siAeJGx9LYSgToOLBzotYxCKKyEymwMmF1IZJwcQqFiUWy6VmNoDl6lLHVBm8omIvXCqZvYaVAtBBLAgZ7T7aC8dPJ5Mnu8+HNjCZhOPH472n4yduOI6FhbN11/kdhm+Apoh6wUyyfi85D8ZDYn9NwpJVkJ4qmsFCUaGXUmUbTm1Ih1wmlPt18rTNOQMUS41Ujmc8EFKZdTwY+tE1aBMPPLluN7SHztw1on2yy7/ZAyEGKrOPuUdSZkeyqmsRIoUGDgnOLQuRuNYFO00SIX7SVQ+amKUQzXNIdPSaGhp57Oji6H1UV2/O4YqbpWXq0yxfe761p9thL7vqJGNasxK2xrBF6F4/hOkHwZnZHk1Xvr0Y5Uriw6AhfQjVc6ENFQmkW6B7hnTeSpqCirpI0YovTqMrWY1YE/oj5u6Z4dQy/XciOqA9ekZmM/23tF0ys54xMaPVf1G5gf+N3lHG8E+rP9J9LgwoDrT8lzdmB7RPI9tM31vbjCX4lEqdbVdNC+P5Z83w3nwXCr+LKw5212VhzkXK8B2zXfL9mF5J/9y9ZV0CNWtrFrapowbxxK/94JdMY/HZfa9bDl3cQvGakbdDoU7QAoS5Yhn6mRKboCCTJbxCp+HLtl96/ND/LIemaa2oNh4O7KTKAdcAeiruPJFVFbTtQCr79jBsAkwkCO6W8xurfoF9LhRo7CflGtqFjs6MGsWq/R6rs75jmbqGp9P4xsINNyF0X51RXfmEc5ZrydLw8mz+bNJZ0OO92i2pd8SdfSM7TU8wRdFwyW8W8kgWwtKd57ixEHSs352Zli/ZHe8Mm9sqLZR7T+5bF2kBEevW27U5FWlCteFgN2ghJb+iagaiCLxz23kxGA6m2txwOGx4v2RZjh7Q7ksQhpGBLMf3MD43V0XyBUyYaN3cbdOomzpNWUlYetBjpUnCqdY4syw4n7OvaLkOpxGu/yGVS7dz70p8M9Ibu2y9e/jWB8MwnEY47M80Xt2dyt8B). It doesn't seem to work.

File: Specs/Data/Models/PBR/BoxUnlit/BoxUnlit-processed.gltf (hunk 19 above)
Comment (ptrgags) [reply Specs/Data/Models/PBR/BoxUnlit/BoxUnlit-processed.gltf:1]: Oh whoops this shouldn't have been committed. Accidental output when using `gltf-pipeline` incorrectly to generate a GLB :sweat_smile: I'll remove this.

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl (hunk 20 above)
Comment (ptrgags) [reply Source/Shaders/ModelExperimental/MaterialStageFS.glsl:109]: Oh okay this was me misunderstanding the `if/elif/else` in GLSL clause, I thought if you had an else clause you can't use `ifdef` but that's not true I'm learning now. You can still use `#ifdef`, there's just no corresponding `#elifdef`, you have to do `#elif defined()`. Here's an example (works in ShaderToy):

```glsl
//#define RED
#define GREEN

void mainImage( out vec4 fragColor, in vec2 fragCoord )
{
    // Normalized pixel coordinates (from 0 to 1)
    vec2 uv = fragCoord/iResolution.xy;

    vec3 color;
    #ifdef RED
    color = vec3(1.0, 0.0, 0.0);
    #elif defined(GREEN)
    color = vec3(0.0, 1.0, 0.0);
    #else
    color= vec3(1.0);
    #endif

    // Output to screen
    fragColor = vec4(color,1.0);
}
```

I'll update the shaders to use `#ifdef` when there's one condition.

**Comment by ptrgags** [comment]
@sanjeetsuhag for the emissive box, what glTF viewer were you using? I'll compare with what the old Model does, when PBR is off does the emissive color get used at all?

Also I just checked the [`KHR_materials_unlit` spec](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_unlit#extending-materials), the spec implies that emissive should be ignored:

>When present, the extension indicates that a material should be unlit and use available baseColor values, alpha values, and vertex colors while ignoring all properties of the default PBR model related to lighting or color. Alpha coverage and doubleSided still apply to unlit materials.

We might want to discuss it more. if we do want emissive textures in unlit mode, then it's just a matter of rearranging some of the material and lighting shader code.

**Comment by ptrgags** [comment]
@sanjeetsuhag I tried modifying BoxEmissive to include `KHR_materials_unlit` and ran it with the old Model. It produces a black box which is consistent with my branch and the glTF spec.

image

So I'm not going to change the shader to use the emissive color

**Comment by ptrgags** [comment]
From the latest build looks like I need to:

- [ ] Update Gltf loader tests given that I added that missing otherwise clause

**Comment by ptrgags** [comment]
@sanjeetsuhag ugh the timeouts are back. I temporarily disabled the material pipeline test just to experiment with how Travis will react

**Comment by ptrgags** [comment]
Swaping up BoxUnlit -> UnlitTest worked? :man_shrugging: I'm restarting the builds to make sure this isn't a fluke

**Comment by sanjeetsuhag** [comment]
@ptrgags This PR looks good to me, the models all loading, tests and coverage look good. If CI passes on both, I'll merge.

**Comment by sanjeetsuhag** [comment]
CI passed, so I'm merging. Thanks @ptrgags!
